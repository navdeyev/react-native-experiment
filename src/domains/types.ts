import {ThunkAction} from 'redux-thunk';
import {ServiceMap} from 'src/domains/services';
import {NavigationAction, NavigationParams} from "react-navigation";

export interface Geo {
    lat: string;
    lng: string;
}

export interface Address {
    street: string;
    suite: string;
    city: string;
    zipcode: string;
    geo: Geo;
}

export interface Company {
    name: string;
    catchPhrase: string;
    bs: string;
}

export interface User {
    id: number;
    name: string;
    username: string;
    email: string;
    address: Address;
    phone: string;
    website: string;
    company: Company;
}

export interface UserDataState {
    userData: Array<User>
}

export interface AppState {
    userDataState: UserDataState
}

export type AppThunkAction<R> = ThunkAction<R, AppState, ServiceMap>

export interface NavigateOptions {
    routeName: string;
    params?: NavigationParams;
    action?: NavigationAction;
    key?: string;
}

export interface HasGetParam<P> {
    getParam: <T extends keyof P>(param: T, fallback?: P[T]) => P[T];
}
