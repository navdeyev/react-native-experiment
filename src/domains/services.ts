import {createUserDataService, UserDataService} from 'src/domains/userData/userDataService';

export interface ServiceMap {
    userDataService: UserDataService
}

export const createServiceMap = () => {
    return {
        userDataService: createUserDataService()
    }
};
