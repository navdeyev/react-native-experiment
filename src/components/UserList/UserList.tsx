import i18n from 'i18n-js';
import * as React from 'react';
import styled from 'styled-components';
import {Button, FlatList, ListRenderItemInfo, Text, View} from 'react-native';

import {User} from '../../domains/types';

interface RowProps {
    lastChild: boolean
}

export const StyledRow = styled(View)`
  padding: 8px;
  display: flex;
  border-radius: 4px;
  flex-direction: row;
  border: 2px solid #333;
  margin-bottom: ${(props: RowProps) => props.lastChild ? '0' : '8px'};
`;

const StyledText = styled(Text)`
  flex: 1;
  padding: 7px 8px 7px 0;
`;

export const renderItem = (onSelectUser: (user: User) => void, dataLength: number) =>
    (info: ListRenderItemInfo<User>) => {
        const user = info.item;
        const pressHandler = () => onSelectUser(user);
        const text = `${ user.username } <${ user.name }>`;
        return (
            <StyledRow lastChild={info.index === dataLength - 1}>
                <StyledText testID={`user-row-text-${info.index}`}>{text}</StyledText>
                <Button title={i18n.t('details')}
                        onPress={pressHandler}
                        testID={`view-details-button-${info.index}`}/>
            </StyledRow>
        );
    };

export interface Props {
    userData: Array<User>;
    onSelectUser: (user: User) => void;
}

export const keyExtractor = (user: User) => '' + user.id;

const UserList: React.SFC<Props> = (props) => {
    const {userData, onSelectUser} = props;
    return (
        <FlatList
            data={userData}
            renderItem={renderItem(onSelectUser, userData.length)}
            keyExtractor={keyExtractor}
        />
    );
};

export default UserList;