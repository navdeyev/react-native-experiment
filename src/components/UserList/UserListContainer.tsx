import * as React from 'react';
import {connect} from 'react-redux';

import {AppState, AppThunkAction, User} from 'src/domains/types';
import UserList from 'src/components/UserList/UserList';
import * as userDataActions from 'src/domains/userData/userDataActions';

interface Props {
    userData: Array<User>,
    loadUserList: () => AppThunkAction<Promise<any>>
}

class UserListContainer extends React.Component<Props> {
    componentDidMount() {
        this.props.loadUserList();
    }

    render() {
        return <UserList userData={this.props.userData}/>;
    }
}

const mapActionCreatorsToProps = {
    loadUserList: userDataActions.loadUserList
};

const mapStateToProps = (state: AppState) => ({
    userData: state.userDataState.userData
});

export default connect(mapStateToProps, mapActionCreatorsToProps)(UserListContainer);