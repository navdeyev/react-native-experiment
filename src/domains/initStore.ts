import {AnyAction, applyMiddleware, createStore, Store} from 'redux';
import thunk from 'redux-thunk';

import rootReducer from 'src/domains/rootReducer';
import {createServiceMap} from 'src/domains/services';
import {AppState} from 'src/domains/types';

const initStore = (): Store<AppState> => {
    const serviceMap = createServiceMap();

    const thunkMiddleware = thunk.withExtraArgument(serviceMap);
    const middlewares = [
        thunkMiddleware
    ];
    const enhancer = applyMiddleware(...middlewares);

    return createStore<AppState, AnyAction, {}, {}>(rootReducer, enhancer);
};

export default initStore;