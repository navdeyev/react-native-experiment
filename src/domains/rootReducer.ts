import {combineReducers} from 'redux';

import {AppState} from 'src/domains/types';
import userDataReducer from 'src/domains/userData/userDataReducer';

const rootReducer = combineReducers<AppState>({
    userDataState: userDataReducer
});

export default rootReducer;