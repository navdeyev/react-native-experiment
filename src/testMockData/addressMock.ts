import {Address, Geo} from 'src/domains/types';

export const geo: Geo = {
    lat: '-37.3159',
    lng: '81.1496'
};

export const address: Address = {
    street: 'Kulas Light',
    suite: 'Apt. 556',
    city: 'Gwenborough',
    zipcode: '92998-3874',
    geo
};
