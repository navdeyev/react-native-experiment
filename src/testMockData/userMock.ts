import {User} from '../domains/types';

import {address} from './addressMock';
import {company} from './companyMock';

export const user: User = {
    id: 1,
    name: 'Leanne Graham',
    username: 'Bret',
    email: 'Sincere@april.biz',
    address,
    phone: '1-770-736-8031 x56442',
    website: 'hildegard.org',
    company
};
