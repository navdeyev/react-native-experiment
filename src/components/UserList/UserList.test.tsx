import 'jest-styled-components';
import * as React from 'react';
import {shallow} from 'enzyme';
import {ListRenderItemInfo} from 'react-native';

import {User} from 'src/domains/types';
import {user} from 'src/testMockData/userMock';

import UserList, {keyExtractor, Props, renderItem, StyledRow} from './UserList';

describe('UserList', () => {

    describe('renderItem', () => {

        let itemInfo: ListRenderItemInfo<User>;
        beforeEach(() => {
            itemInfo = {
                item: user,
                index: 0,
                separators: {
                    highlight: jest.fn(),
                    unhighlight: jest.fn(),
                    updateProps: jest.fn(),
                }
            };
        });

        it('renders the component', () => {
            const onSelectUser = jest.fn();
            const rendered = shallow(renderItem(onSelectUser, 10)(itemInfo));
            expect(rendered.prop('lastChild')).toBe(false);

            const userLabel = rendered.find('[testID="user-row-text-0"]');
            expect(userLabel.prop('children')).toEqual('Bret <Leanne Graham>');

            expect(rendered.getElement()).toMatchSnapshot();
        });

        it('renders the component as the last child', () => {
            const onSelectUser = jest.fn();
            const rendered = shallow(renderItem(onSelectUser, 1)(itemInfo));
            expect(rendered.prop('lastChild')).toBe(true);
            expect(rendered.getElement()).toMatchSnapshot();
        });

        it('executes onSelectUser when row is pressed', () => {
            const onSelectUser = jest.fn();
            const rendered = shallow(renderItem(onSelectUser, 1)(itemInfo));
            const highlight = rendered.find('[testID="view-details-button-0"]');
            highlight.simulate('press');
            expect(onSelectUser).toHaveBeenCalledWith(itemInfo.item);
        });
    });

    describe('UserList', () => {
        it('renders the component', () => {
            const props: Props = {
                userData: [user],
                onSelectUser: jest.fn()
            };
            const rendered = shallow(<UserList {...props} />);
            expect(rendered.getElement()).toMatchSnapshot();
        });
    });

    describe('keyExtractor', () => {
        it('generates a key', () => {
            expect(keyExtractor(user)).toBe('' + user.id);
        });
    });
});

describe('StyledRow', () => {
    it('adds bottom margin to the component if it is not the last child', () => {
        const rendered = shallow(<StyledRow lastChild={false}/>);
        expect(rendered).toHaveStyleRule('margin-bottom', '8px');
    });

    it('adds removes bottom margin from the component if it is the last child', () => {
        const rendered = shallow(<StyledRow lastChild={true}/>);
        expect(rendered).toHaveStyleRule('margin-bottom', '0');
    });
});
