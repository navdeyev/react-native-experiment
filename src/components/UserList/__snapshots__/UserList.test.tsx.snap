// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList UserList renders the component 1`] = `
<FlatList
  data={
    Array [
      Object {
        "address": Object {
          "city": "Gwenborough",
          "geo": Object {
            "lat": "-37.3159",
            "lng": "81.1496",
          },
          "street": "Kulas Light",
          "suite": "Apt. 556",
          "zipcode": "92998-3874",
        },
        "company": Object {
          "bs": "harness real-time e-markets",
          "catchPhrase": "Multi-layered client-server neural-net",
          "name": "Romaguera-Crona",
        },
        "email": "Sincere@april.biz",
        "id": 1,
        "name": "Leanne Graham",
        "phone": "1-770-736-8031 x56442",
        "username": "Bret",
        "website": "hildegard.org",
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;

exports[`UserList renderItem renders the component 1`] = `
<Component
  className="sc-bdVaJa dOJfCv"
  lastChild={false}
>
  <Styled(Text)
    testID="user-row-text-0"
  >
    Bret &lt;Leanne Graham&gt;
  </Styled(Text)>
  <Button
    onPress={[Function]}
    testID="view-details-button-0"
    title="Details"
  />
</Component>
`;

exports[`UserList renderItem renders the component as the last child 1`] = `
<Component
  className="sc-bdVaJa hRZBlw"
  lastChild={true}
>
  <Styled(Text)
    testID="user-row-text-0"
  >
    Bret &lt;Leanne Graham&gt;
  </Styled(Text)>
  <Button
    onPress={[Function]}
    testID="view-details-button-0"
    title="Details"
  />
</Component>
`;
